File: app/src/main/java/tianyinews/tianyi/com/tianyinews/fragment/childfragment/HomeChildFragment.java
package tianyinews.tianyi.com.tianyinews.fragment.childfragment;

import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.view.View;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.recyclerview.widget.RecyclerView;
import com.loopj.android.http.AsyncHttpClient;
import com.loopj.android.http.TextHttpResponseHandler;
import com.scwang.smart.refresh.layout.SmartRefreshLayout;
import com.scwang.smart.refresh.layout.api.RefreshLayout;
import com.scwang.smart.refresh.layout.listener.OnLoadMoreListener;
import com.scwang.smart.refresh.layout.listener.OnRefreshListener;
import java.util.ArrayList;
import java.util.List;
import cz.msebera.android.httpclient.Header;
import tianyinews.tianyi.com.tianyinews.R;
import tianyinews.tianyi.com.tianyinews.activity.WebActivity;
import tianyinews.tianyi.com.tianyinews.adapter.MyHomeListViewAdapter;
import tianyinews.tianyi.com.tianyinews.base.BaseFragment;
import tianyinews.tianyi.com.tianyinews.bean.MyChannel;
import tianyinews.tianyi.com.tianyinews.bean.NewsBean;
import tianyinews.tianyi.com.tianyinews.util.JsonUtil;

/**
 * @类的用途:
 * @作者: 任正威
 * @date: 2017/3/14.
 */

public class HomeChildFragment extends BaseFragment {
    public static final java.lang.String MODEL_KEY = "model_key";
    private int PageIndex = 1;
    private MyHomeListViewAdapter adapter;
    private MyChannel mChannel;
    private RecyclerView recyclerView;
    private SmartRefreshLayout refreshLayout;

    @Override
    protected int getLayoutId() {
        return R.layout.homechildfragment;
    }

    @Override
    protected void initView(View view) {
        refreshLayout = view.findViewById(R.id.refreshLayout);
        recyclerView = view.findViewById(R.id.recyclerView);
        Bundle bundle = getArguments();
        mChannel = (MyChannel) bundle.getSerializable(MODEL_KEY);
        adapter = new MyHomeListViewAdapter();
        recyclerView.setAdapter(adapter);
    }

    @Override
    protected void initData() {
        super.initData();
        //获取网络数据
        getServerData(true);

    }



    @Override
    protected void initListener() {
        super.initListener();
        refreshLayout.setOnRefreshListener(new OnRefreshListener() {
            @Override
            public void onRefresh(@NonNull RefreshLayout refreshLayout) {
                PageIndex = 1;
                getServerData(true);
            }
        });

        refreshLayout.setOnLoadMoreListener(new OnLoadMoreListener() {
            @Override
            public void onLoadMore(@NonNull RefreshLayout refreshLayout) {
                PageIndex++;
                getServerData(false);
            }
        });

        adapter.setOnItemClickListener((ada, view, position) -> {
            NewsBean.ResultBean.DataBean item = adapter.getItem(position);
            if(item==null)return;
            String newUrl = item.url;
            Intent intent = new Intent(getActivity(), WebActivity.class);
            intent.putExtra("newUrl", newUrl);
            startActivity(intent);
        });

        adapter.setOnItemLongClickListener((adapter, view, position) -> {
            AlertDialog.Builder builder = new AlertDialog.Builder(mContext);
            builder.setTitle("收藏精彩看点");
            builder.setMessage("确定收藏吗？收藏以后更方便阅读精彩内容");
            builder.setPositiveButton("确定", new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialogInterface, int i) {
                    //收藏状态是在已经登录的情况下才能收藏
                    //收藏到数据库并将关注页面状态更改
                }
            });
            builder.setNegativeButton("取消", (dialogInterface, i) -> dialogInterface.dismiss());
            builder.show();

            return false  ;
        });
    }

    private void getServerData(boolean isResresh) {
        AsyncHttpClient asyncHttpClient = new AsyncHttpClient();
        String http_url = mChannel.url + "?" + "type=" + mChannel.requestTitle + "&" + "page=" + PageIndex + "&" + mChannel.url_footer;
        //   Log.e("sadddddddddd", http_url);
        asyncHttpClient.get(getContext(), http_url, new TextHttpResponseHandler() {
            @Override
            public void onFailure(int statusCode, Header[] headers, String responseString, Throwable throwable) {
                String url = "";
            }

            @Override
            public void onSuccess(int statusCode, Header[] headers, String responseString) {
                List<NewsBean.ResultBean.DataBean> data = JsonUtil.getJson(responseString);
                if(isResresh){
                    adapter.setList(data);
                    refreshLayout.finishRefresh();
                }else{
                    adapter.addData(data);
                    refreshLayout.finishLoadMore();
                }
            }
        });
    }

    public String getTitle() {
        return mChannel.title;
    }

}

File: app/src/main/java/tianyinews/tianyi/com/tianyinews/fragment/childfragment/HomeChildFragment.kt
package tianyinews.tianyi.com.tianyinews.fragment.childfragment

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.widget.TextView
import androidx.lifecycle.Observer
import com.chad.library.adapter.base.BaseQuickAdapter
import com.kingja.loadsir.core.LoadService
import kotlinx.android.synthetic.main.homechildfragment.*
import tianyinews.tianyi.com.tianyinews.R
import tianyinews.tianyi.com.tianyinews.activity.WebActivity
import tianyinews.tianyi.com.tianyinews.adapter.MyHomeListViewAdapter
import tianyinews.tianyi.com.tianyinews.base.BaseFragment
import tianyinews.tianyi.com.tianyinews.bean.MyChannel
import tianyinews.tianyi.com.tianyinews.bean.NewsDataModel
import tianyinews.tianyi.com.tianyinews.databinding.HomechildfragmentBinding
import tianyinews.tianyi.com.tianyinews.ext.loadServiceInit
import tianyinews.tianyi.com.tianyinews.ext.showLoading
import tianyinews.tianyi.com.tianyinews.ext.showMessage
import tianyinews.tianyi.com.tianyinews.viewmodel.HomeChildViewModel
import java.util.*

/**
 * @类的用途:
 * @作者: 任正威
 * @date: 2017/3/14.
 */
class HomeChildFragment : BaseFragment<HomeChildViewModel, HomechildfragmentBinding>() {

    private val adapter: MyHomeListViewAdapter by lazy { MyHomeListViewAdapter() }
    private var mChannel: MyChannel? = null
    private val headerView: View by lazy { LayoutInflater.from(requireContext()).inflate(R.layout.header_message_update, null) }
    private val tv_title: TextView by lazy { headerView.findViewById(R.id.tv_title) }

    //界面状态管理者
    private val loadsir: LoadService<Any> by lazy {
        loadServiceInit(mDatabind.refreshLayout) {
            //点击重试时触发的操作
            loadsir.showLoading()
            mViewModel.reuqestData(true, mChannel!!.requestTitle)
        }
    }

    override fun layoutId(): Int = R.layout.homechildfragment

    override fun initView(savedInstanceState: Bundle?) {
        val bundle = arguments
        mChannel = bundle!!.getSerializable(MODEL_KEY) as MyChannel?
        mDatabind.recyclerView.adapter = adapter
        adapter.setAnimationWithDefault(BaseQuickAdapter.AnimationType.AlphaIn)
    }

    override fun lazyLoadData() {
        //设置界面 加载中
        loadsir.showLoading()
        //获取网络数据
        mViewModel.reuqestData(true, mChannel!!.requestTitle)
    }

    override fun initListener() {
        mDatabind.refreshLayout?.setOnRefreshListener {
            mViewModel.reuqestData(true, mChannel!!.requestTitle)
        }
        mDatabind.refreshLayout?.setOnLoadMoreListener {
            mViewModel.reuqestData(false, mChannel!!.requestTitle)
        }
        adapter.setOnItemClickListener { ada: BaseQuickAdapter<*, *>, view: View?, position: Int ->
            val item = adapter.getItem(position) ?: return@setOnItemClickListener
            val newUrl = item.url
            WebActivity.start(requireContext(), newUrl)
        }
        adapter.setOnItemLongClickListener { adapter: BaseQuickAdapter<*, *>, view: View?, position: Int ->
            showMessage("确定收藏吗？收藏以后更方便阅读精彩内容", "收藏精彩看点", positiveAction = {
                //收藏状态是在已经登录的情况下才能收藏
                //收藏到数据库并将关注页面状态更改
            }, negativeAction = {
                dismissLoading()
            })
            false
        }
    }

    private var insertSpace = 0 // 广告间距
    private var lastInsertIndex = 0 // 广告最后插入位置

    /**
     * 插入广告数据
     */
    private fun integrateAdToNewsData(list: List<NewsDataModel>?):MutableList<NewsDataModel> {
        var newsInfoModelList = mutableListOf<NewsDataModel>()
        try {
            if (list == null) return mutableListOf()
            newsInfoModelList =  list.toMutableList()
            if (insertSpace == 0) {
                insertSpace = 6
            }
            while (lastInsertIndex + insertSpace < newsInfoModelList.size) {
                val newsInfoModel = NewsDataModel()
                newsInfoModel.uniquekey = "ad"
                lastInsertIndex += this.insertSpace
                newsInfoModelList.add(lastInsertIndex, newsInfoModel)
                lastInsertIndex++
                if (insertSpace == 6) {
                    insertSpace = 9
                } else if (insertSpace == 9) {
                    insertSpace = 9
                }
            }
            lastInsertIndex -= newsInfoModelList.size
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return newsInfoModelList
    }

    override fun createObserver() {
        mViewModel.run {
            homeDataState.observe(viewLifecycleOwner, Observer {
                if (it.isRefresh) {
                    adapter.setList(integrateAdToNewsData(it.listData))
                    loadsir.showSuccess()
                    refreshLayout!!.finishRefresh()
                    if (adapter.hasHeaderLayout()) {
                        adapter.removeAllHeaderView()
                    }
                    adapter.addHeaderView(headerView)
                    tv_title.text = "今日头条推荐引擎有${it.listData?.size}条更新"
                    tv_title.postDelayed({
                        adapter.removeHeaderView(headerView)
                    }, 3000)
                } else {
                    it.listData?.let {
                        adapter.addData(integrateAdToNewsData(it))
                        loadsir.showSuccess()
                    }

                    mDatabind.refreshLayout.finishLoadMore()
                }

            })
        }
    }

    val title: String
        get() = mChannel!!.title

    companion object {
        const val MODEL_KEY = "model_key"

        @JvmStatic
        fun newInstance(channel: MyChannel): HomeChildFragment {
            val args = Bundle()
            args.putSerializable(MODEL_KEY, channel)
            val fragment = HomeChildFragment()
            fragment.arguments = args
            return fragment
        }
    }


}